src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,143 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public boolean add(T value) {
-        return false;
+        Node<T> current = new Node<T>(tail, value, null);
+        if (size == 0) {
+            head = current;
+        } else {
+            tail.next = current;
+        }
+        tail = current;
+        size++;
+        return true;
+
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index == size) {
+            add(value);
+            return;
+        }
+        checkElementIndex(index);
+        Node<T> oldNode = getNode(index);
+        Node<T> newNode = new Node<>(oldNode.previous, value, oldNode);
+        if (index > 0) {
+            newNode.previous.next = newNode;
+        } else {
+            head = newNode;
+        }
+        oldNode.previous = newNode;
+        size++;
     }
 
     @Override
     public boolean addAll(List<T> list) {
-        return false;
+        for (T value : list) {
+            add(value);
+        }
+        return true;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkElementIndex(index);
+        return getNode(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = getNode(index);
+        T oldVal = node.element;
+        node.element = value;
+        return oldVal;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkElementIndex(index);
+        Node<T> node = getNode(index);
+        if (index > 0) {
+            unlink(node);
+        } else {
+            head = node.next;
+            size--;
+        }
+        return node.element;
     }
 
     @Override
     public boolean remove(T t) {
+        Node<T> node = head;
+        for (int i = 0; i < size; i++) {
+            if (node.element == t || node.element.equals(t)) {
```

Comment:
what does happen if node.element == null?

Reply:
should return true, done